### ISSUE TITLE
Super Stardust Delta [PCSA00006]

### ISSUE THREAD

**Issue by arnoldsecret** [Vita3K/compatibility#441]
# App summary
- App name: Super Stardust Delta
- App serial: PCSA00006
- App version: 1.00

# Vita3K summary
- Version: v0.1
- Build number: 1822
- Commit hash: https://github.com/Vita3K/Vita3K/commit/4938ca8b164147e462e01af7eec4ec3e0a475101 

# Test environment summary
- Tested by: Arnold Secret
- OS: Windows 10
- CPU: Intel Core i7-4500U
- GPU: NVIDIA Geforce GT 740M
- RAM: 8 GB

# Issues
Unicorn error on initial loading.

# Screenshots
![Untitled](https://user-images.githubusercontent.com/41223243/104275149-43630300-54dd-11eb-9615-f7554e61481d.png)

# Log
[vita3k.log](https://github.com/Vita3K/compatibility/files/5800006/vita3k.log)

**Comment by Zangetsu38** [comment]
not need screenshots if is locked on loading screen :)

**Comment by GamingDucking** [comment]
# Update (either Ingame or Playable)
On https://github.com/Vita3K/Vita3K/commit/f803e6152024b97276c81a258011517de10da6bc the game now runs ingame at 24 fps.
![Screenshot 2022-06-01 123049 PNG](https://user-images.githubusercontent.com/90236481/171375760-ca64d820-77e0-4c79-a7a6-2f4ac5281716.jpg)
![Screenshot 2022-06-01 123158](https://user-images.githubusercontent.com/90236481/171375773-19601df1-2072-47bd-af59-3cc504946022.jpg)
![Screenshot 2022-06-01 123416](https://user-images.githubusercontent.com/90236481/171375789-0b9a7395-99cc-4653-8b5d-a10a28f59f78.jpg)
# Logs
[vita3k.log](https://github.com/Vita3K/compatibility/files/8812987/vita3k.log)

**Comment by ViRGEdx** [comment]
Kudos to the devs on their progress. Compatibility on this game is looking noticeably better as of v0.2.03688. (Core i9 9900K + RTX 2080)

**OpenGL**: the render flashing that made the game effectively unplayable has been resolved. As a result there are no longer immediate "breaking" graphical glitches in the game. However, there are still remaining nuisance graphical glitches, most notably, phantom effects from weapons being drawn in areas they shouldn't be (only with weapons aimed up and/or left, strangely enough). Performance also remains sub-optimal, with framerates regularly dropping as low as the 30s..

![Vita3K_Stardust](https://github.com/user-attachments/assets/ba7fc0aa-3c40-49b0-9528-0629ec3e675c)

**Vulkan**: The game still crashes while trying to load an actual stage with a *vk::Queue::submit: ErrorDeviceLost* error. But it's a solid 60fps right up to the point it crashes!

**Comment by amurgshere** [comment]
Just opened PR [#4139](https://github.com/Vita3K/Vita3K/pull/4139) which should fix the three issues described above:

- The phantom weapon effects being drawn in the wrong place (the glow appearing near the HUD icon, and a fainter version at the top of the screen on OpenGL) turned out to be two related bugs in how GXM downscale color surfaces are handled, one affecting both backends and one OpenGL only.
- The missing green stardust particles on OpenGL are fixed too. `GL_PROGRAM_POINT_SIZE` was never enabled, so shader-written point sizes were ignored and the point-sprite particles didn't show. This was OpenGL only, it never occurred on Vulkan.
- The Vulkan crash (ErrorDeviceLost from vkQueueSubmit) was caused by incorrect render target size tracking that could request a render area larger than the actual framebuffer.

These are all fixes to shared, platform agnostic rendering code, so they should apply equally on Android, macOS and Linux, not just Windows.

Testing scope: I've only actually tested this against Super Stardust Delta on Windows, on both Vulkan and OpenGL. I have not tested on Android, macOS, or Linux, so treat the fix as verified on Windows and expected elsewhere rather than confirmed everywhere. The framerate issues mentioned above are not addressed by this PR, they look like a separate problem.

EDIT: I have now tested this on Android too and it works well, though my Retroid Pocket 5 had additional issues with the system Qualcomm GPU driver crashing, i have additional fixes targeted at that chip coming in a separate PR.

I'm also separately working on fixing a crash in this game on Android with the system Qualcomm drivers, that will come as its own PR, afterall its such a great game!

<img width="962" height="576" alt="Image" src="https://github.com/user-attachments/assets/b2036a95-b777-4aa9-9e10-5b7f6434ab6a" />

**Comment by ViRGEdx** [comment]
@amurgshere, thank you for the hard work you put into fixing this game. I've tried out the Windows artifact build of your pull request, and the game is working fantastically. I've not had a chance to exhaustively check every last feature, but baring any bugs in a niche feature, this game should be suitable for promotion to playable status once the PR is merged.

Vulkan performance has been a rock-solid 60fps. OpenGL less so, but it's pretty common for that render backend to underperform.

(This is the best version of Super Stardust, so I am super excited that it's finally playable in an emulator! Being that it's a handheld port(ish) of a PS3 game, I wasn't expecting anyone to get around to this for quite a while.)

**Comment by amurgshere** [comment]
Thank you @ViRGEdx , I appreciate you saying that. Note that I do have a fix for the OpenGL performance in this game (and it should improve performance in many other games too). Large textures were being uploaded to the GPU synchronously, stalling the graphics thread and causing laggy performance. It now stages those uploads through a pixel buffer object so the driver can transfer them without holding up the graphics thread (I saw a smooth 60 fps on OpenGL on my Windows machine after these changes in Super Stardust Delta from a 52-58 sporadic framerate). See my PR here for details:

https://github.com/Vita3K/Vita3K/pull/4161

**Comment by gerd3000** [comment]
> [@amurgshere](https://github.com/amurgshere), thank you for the hard work you put into fixing this game. I've tried out the Windows artifact build of your pull request, and the game is working fantastically. I've not had a chance to exhaustively check every last feature, but baring any bugs in a niche feature, this game should be suitable for promotion to playable status once the PR is merged.
> 
> Vulkan performance has been a rock-solid 60fps. OpenGL less so, but it's pretty common for that render backend to underperform.
> 
> (This is the best version of Super Stardust, so I am super excited that it's finally playable in an emulator! Being that it's a handheld port(ish) of a PS3 game, I wasn't expecting anyone to get around to this for quite a while.)



i 100% agree!!!

**Comment by gerd3000** [comment]
this is one of my favorite games of all time,it never gets old and has the best mechanics and game design of all star dust games (havent tried ps4 ultra version though).just pure focus on the basics that make this game so long lasting addictive fun with simple and tight controls. so i too would love to play this on my i5 1335u 64gb dual channel setup one day with the perfect and enhanced performance this game deserves. i have two whitelisted pstvs where i need to get to work all the features, esp. endless mode. cheers
